framework/src/main/java/org/checkerframework/framework/type/DefaultTypeHierarchy.java

```diff
@@ -988,46 +1075,57 @@ protected boolean visitTypevarSupertype(
    * upper bound against the supertypes lower bound.
    */
   protected boolean visitTypevarSubtype(
       AnnotatedTypeVariable subtype, AnnotatedTypeMirror supertype) {
-    AnnotatedTypeMirror upperBound = subtype.getUpperBound();
-    if (TypesUtils.isBoxedPrimitive(upperBound.getUnderlyingType())
+    AnnotatedTypeMirror subtypeUpperBound = subtype.getUpperBound();
+    if (TypesUtils.isBoxedPrimitive(subtypeUpperBound.getUnderlyingType())
         && supertype instanceof AnnotatedPrimitiveType) {
-      upperBound = supertype.atypeFactory.getUnboxedType((AnnotatedDeclaredType) upperBound);
+      subtypeUpperBound =
+          supertype.atypeFactory.getUnboxedType((AnnotatedDeclaredType) subtypeUpperBound);
     }
     if (supertype.getKind() == TypeKind.DECLARED
         && TypesUtils.getTypeElement(supertype.getUnderlyingType()).getKind()
             == ElementKind.INTERFACE) {
-      // Make sure the upper bound is no wildcard or type variable
-      while (upperBound.getKind() == TypeKind.TYPEVAR
-          || upperBound.getKind() == TypeKind.WILDCARD) {
-        if (upperBound.getKind() == TypeKind.TYPEVAR) {
-          upperBound = ((AnnotatedTypeVariable) upperBound).getUpperBound();
+      // Make sure the upper bound is no wildcard or type variable.
+      while (subtypeUpperBound.getKind() == TypeKind.TYPEVAR
+          || subtypeUpperBound.getKind() == TypeKind.WILDCARD) {
+        if (subtypeUpperBound.getKind() == TypeKind.TYPEVAR) {
+          subtypeUpperBound = ((AnnotatedTypeVariable) subtypeUpperBound).getUpperBound();
         }
-        if (upperBound.getKind() == TypeKind.WILDCARD) {
-          upperBound = ((AnnotatedWildcardType) upperBound).getExtendsBound();
+        if (subtypeUpperBound.getKind() == TypeKind.WILDCARD) {
+          subtypeUpperBound = ((AnnotatedWildcardType) subtypeUpperBound).getExtendsBound();
         }
       }
       // If the supertype is an interface, only compare the primary annotations.
       // The actual type argument could implement the interface and the bound of
       // the type variable must not implement the interface.
-      if (upperBound.getKind() == TypeKind.INTERSECTION) {
+      if (subtypeUpperBound.getKind() == TypeKind.INTERSECTION) {
         Types types = checker.getTypeUtils();
-        for (AnnotatedTypeMirror ub : ((AnnotatedIntersectionType) upperBound).getBounds()) {
+        for (AnnotatedTypeMirror bound :
+            ((AnnotatedIntersectionType) subtypeUpperBound).getBounds()) {
+          // Make sure the upper bound is no wildcard or type variable.
+          while (bound.getKind() == TypeKind.TYPEVAR || bound.getKind() == TypeKind.WILDCARD) {
+            if (bound.getKind() == TypeKind.TYPEVAR) {
+              bound = ((AnnotatedTypeVariable) bound).getUpperBound();
+            }
+            if (bound.getKind() == TypeKind.WILDCARD) {
+              bound = ((AnnotatedWildcardType) bound).getExtendsBound();
+            }
+          }
```

Comment:
This code appears identically above (but for a variable name).  I think the method will be easier to read and understand  if you abstract out this logic into a method of its own.